Openstack cinder v1/v2/auto API support

mkutsevol opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
It adds support for v2 cinder API + autodetection of available cinder API level (as in LBs).
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #39572

**Special notes for your reviewer**:
Based on work by @anguslees. The first two commits are just rebased from https://github.com/kubernetes/kubernetes/pull/36344 which already had a lgtm by @jbeda 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```
Add support for v2 cinder API for openstack cloud provider. By default it autodetects the available version.
```

mkutsevol committed:
Support for v1/v2/autoprobe openstack cinder blockstorage

Support for cinder v1/v2 api with the new gophercloud/gophercloud
library. API version is configurable and defaulting autodetection.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40423)
<!-- Reviewable:end -->

xrl commented:
My team is migrating to Mitaka and will lose Cinder V1 support. We are running this code and it's been functional and stable.

xrl commented:
Can I request a bump in the gophercloud version? I am having issues with expiring tokens in my build and a gophercloud update might help.

mkutsevol commented:
Sure @xrl, why not. Will update and push it out tomorrow.

xsgordon commented:
@idvoretskyi can you add the sig/openstack label?

xrl commented:
@jbeda @bgrant0607 can this work be reviewed and merged? it's significant and important to continue support with modern openstack installations.

bgrant0607 commented:
@mkutsevol @xrl This PR needs to be reviewed by someone familiar with Openstack.

@xsgordon @idvoretskyi Please find someone to review.

idvoretskyi commented:
cc @Quentin-M @anguslees

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -524,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
```
This return is suspicious. If v1 API is not enabled, we should keep trying v2.

mkutsevol replied:
And we will, creating a ServiceClient doesn't depend on supported blockstorage api versions

jrperritt replied:
This seems fine: either the user explicitly set "v1" in their config (in which case it should error) or it was determined that "v1" was current.

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -524,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+			return nil, err
+		}
+
+		apiversions_v1.List(sClient).EachPage(func(page pagination.Page) (bool, error) {
```
what if block v1 api is disabled?

mkutsevol replied:
There is no v2 apiversions at all in the gophercloud library.
It is used _only_ to probe available versions. 
So it will work flawlessly if v1 is absent.

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -524,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+			return true, nil
+		})
+		// Autoprobing failed, no supported versions available
+		if os.bsOpts.BSVersion == "auto" {
```
this is redundant, the statement is already under `auto` case.

mkutsevol replied:
It is not, it will be (if probe is successful) switched to highest supported and available version on lines 589/592,
If it is left in "auto" that means that probe failed and it errors out with a corresponding error message.

rootfs commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -23,20 +23,181 @@ import (
+	return vol.ID, err
+}
+
+func (volumes *VolumesV2) Create(opts VolumeCreateOpts) (string, error) {
```
It looks to me, except in `getVolume` where `ServerID` and `Device` are new in V2, `CreateDisk`, `DeleteDisk` and rest of `getVolume` are identical in both V1 and V2. 

In such case, why not just pass version as a new parameter into these functions and call the right package based on version, rather than copying code?

mkutsevol replied:
I tend to agree that duplicate code is not good, but in this particular case, please consider why I did that way. I implement the IVolumes interface. Now, the Create/Delete methods are similar. But when v3 arrives that might not be the case. If I'll have a structure like:
```
func (volumes *VolumesV1) Create(opts VolumeCreateOpts) (string, error) {
	return generalCreate('v1',...)	
}
func (volumes *VolumesV2) Create(opts VolumeCreateOpts) (string, error) {
	return generalCreate('v2',...)	
}
````
... etc

I may force a developer, who will implement v3 support to adapt to that style, and no one knows what the lib will be like at that time. It may be not convenient.

I think that code organized this way (for this particular case) is more manageable. 

When v3 arrives IVolumes implementations may be split into separate files, eg `openstack_volumes_v#`

Did I make you change your opinion? :D

rootfs replied:
I looked at [v3](https://developer.openstack.org/api-ref/block-storage/v3/index.html?expanded=create-a-volume-detail) and [v2](https://developer.openstack.org/api-ref/block-storage/v2/index.html?expanded=create-volume-detail) create volume APIs, so far they look the same.

But I buy your points that things could diverge later.

rootfs commented:
@mkutsevol are you still working on it?

Has this patch been tested on a setup that disables volume v1?

xrl commented:
@rootfs I have been using this PR with a `mitaka` version of openstack (where cinder v1 is gone). It has been working reliably.

rootfs commented:
@xrl thanks for the confirmation.

rootfs commented:
@xrl how do you set `bs-version` option in your cloud config?

xrl commented:
@rootfs I stopped setting bs-version:

```
[Global]
auth-url={{ openstack_auth_url }}
username={{ openstack_username }}
password={{ openstack_password }}
region={{ openstack_region }}
tenant-id={{ openstack_tenant_id }}
domain-name={{ openstack_domain_name }}
[BlockStorage]
# bs-version = v2
trust-device-path = no
```

mkutsevol commented:
I'm here, just been busy. Will reply to your CR :D

mkutsevol commented:
And thank you for the review @rootfs!

rootfs commented:
@mkutsevol please rebase

rootfs commented:
/assign

mkutsevol commented:
@rootfs, will do right now

rootfs commented:
@k8s-bot ok to test

mkutsevol commented:
@anguslees' code was merged meanwhile, so I'll just throw out two first commits, they were a cherrypick of his work.
There will be 2 commits only: code / dep update

mkutsevol committed:
Update deps

mkutsevol commented:
@rootfs, I've rebased (twice :D) and fixed a typo in tests

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -513,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+				return false, err
+			}
+
+			sort.Sort(APIVersionsByID(apiversions))
```
Why are the API versions being sorted? There should be only one "CURRENT" version.

mkutsevol replied:
To get the biggest supported version. If current is v3, but we don't supported it will choose v2 over v1

mkutsevol replied:
It was missing a piece of code down the line, and was unused, you were right. Now it's used.

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -513,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+			return nil, err
+		}
+
+		apiversions_v1.List(sClient).EachPage(func(page pagination.Page) (bool, error) {
```
`EachPage` returns an `error` that should at least be acknowledged, but preferably handled.

mkutsevol replied:
True, will update.

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -513,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+					}
+				}
+			}
+			return true, nil
```
```go
return false, nil
```

mkutsevol replied:
Didn't get it. It succeeded

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -23,20 +23,181 @@ import (
 	"github.com/golang/glog"
 )
 
+type IVolumes interface {
```
Naming is hard; however, you may want to consider changing the name of this interface to be more idiomatic (maybe `VolumeService` with `CreateVolume` and `DeleteVolume`).

mkutsevol replied:
I will just believe you at this point.

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -513,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+	return apiVersions[i].ID > apiVersions[j].ID
+}
+
+func (os *OpenStack) Volumes() (IVolumes, error) {
```
You may want to consider changing the name of this method. Currently, it reads like volumes (an abstract representation of a set of block-storage devices) are getting returned.

xrl commented:
I'm very excited to get off my custom version of kubernetes and start using this on my 30 node cluster!

jrperritt reviewed (changes requested):
I've added a few comments after looking it over. I think adding an interface for different service versions is fine; I also think just type-switching on the version would be fine.

mkutsevol commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -87,7 +89,8 @@ type LoadBalancerOpts struct {
 
 type BlockStorageOpts struct {
-	TrustDevicePath bool `gcfg:"trust-device-path"` // See Issue #33128
+	BSVersion       string `gcfg:"bs-version"`        // overrides autodetection. v1 or v2. Defaults to auto
```
This was done to match already existing code. See loadbalancer's LoadBalancerOpts

mkutsevol commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -513,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
+		}
+		return (IVolumes)(&VolumesV1{sClient, os.bsOpts}), nil
```
True, will update. Good catch.

mkutsevol commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -513,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+			// take first "CURRENT" version
+			for _, version := range apiversions {
+				if version.Status == "CURRENT" {
+					switch version.ID {
```
It's missing a piece of code, it should continue to seek for supported version. Will update.

mkutsevol commented:
@jrperritt , thank  you for the review, I've push a new commit.

rootfs commented:
/lgtm

jrperritt commented:
Aside from a few naming suggestions (I would still change the name of `openstack.Volumes` and drop the 'I' in `IVolumeService`), this looks fine to me.

@mkutsevol alternatively, to avoid committing to a name, those ^ names could just be lower-cased (they're currently only used in the `openstack` package, right?)

mkutsevol commented:
@jrperritt, done and pushed

mkutsevol commented:
@k8s-bot unit test this

rootfs commented:
/lgtm

mkutsevol commented:
@rootfs, @jrperritt what else can I do to get this merged?

rootfs commented:
@wojtek-t can you approve? thanks

rootfs commented:
/lgtm cancel

rootfs commented:
I got a setup to test this patch and got into trouble. 

- Disable block storage v1 and enable just v2:

```console
/etc/cinder/api-paste.ini
Disabled api_v1 in /etc/cinder/api-paste.ini and /etc/cinder/cinder.conf

use = call:cinder.api:root_app_factory
/: apiversions
# /v1: openstack_volume_api_v1
/v2: openstack_volume_api_v2

/etc/cinder/cinder.conf
#enable_v1_api = true
```
-  Creating cinder volume via cli worked

- Clone your branch
```console
# git describe
v1.5.0-alpha.1-8269-g0388a5b
```
- Nothing set in cloud config to test auto detect. 

-  Creating PVC through storage class failed with the following stack trace.

```text
I0316 12:58:07.228083   12759 openstack.go:624] Blockstorage API version probing decided v1
E0316 12:58:07.232098   12759 goroutinemap.go:164] Operation for "provision-default/claim1[33429d46-0a48-11e7-b1fb-fa163ea9f205]" failed. No retries permitted until 2017-03-16 12:58:07.732084907 +0000 UTC (durationBeforeRetry 500ms). Error: recovered from panic "runtime error: invalid memory address or nil pointer dereference". (err=<nil>) Call stack:
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:153
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:290
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/openstack/openstack_volumes.go:88
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/openstack/openstack_volumes.go:296
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder_util.go:168
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder.go:485
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go:1309
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go:1211
/home/fedora/upstream/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/goroutinemap/goroutinemap.go:128
/usr/local/go/src/runtime/asm_amd64.s:2197
```

rootfs commented:
cloned the branch for testing
```bash
#kubectl version
Client Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.1.8269+0388a5b7703051", GitCommit:"0388a5b77030511d4d122b47f0b43d5f523e9ece", GitTreeState:"clean", BuildDate:"2017-03-13T18:16:35Z", GoVersion:"go1.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.1.8269+0388a5b7703051-dirty", GitCommit:"0388a5b77030511d4d122b47f0b43d5f523e9ece", GitTreeState:"dirty", BuildDate:"2017-03-16T13:08:24Z", GoVersion:"go1.8", Compiler:"gc", Platform:"linux/amd64"}
```

But it failed on my setup with the following panic


```
I0316 13:10:58.874055   17018 openstack.go:624] Blockstorage API version probing decided v1
E0316 13:10:58.879364   17018 goroutinemap.go:164] Operation for "provision-default/claim1[ff321889-0a49-11e7-a434-fa163ea9f205]" failed. No retries permitted until 2017-03-16 13:10:59.379352321 +0000 UTC (durationBeforeRetry 500ms). Error: recovered from panic "runtime error: invalid memory address or nil pointer dereference". (err=<nil>) Call stack:
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:153
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:290
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/openstack/openstack_volumes.go:88
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/openstack/openstack_volumes.go:296
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder_util.go:168
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder.go:485
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go:1309
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go:1211
/home/fedora/test/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/goroutinemap/goroutinemap.go:128
/usr/local/go/src/runtime/asm_amd64.s:2197
```

I have disabled block storage v1
```console
/etc/cinder/api-paste.ini
Disabled api_v1 in /etc/cinder/api-paste.ini and /etc/cinder/cinder.conf

use = call:cinder.api:root_app_factory
/: apiversions
# /v1: openstack_volume_api_v1
/v2: openstack_volume_api_v2

/etc/cinder/cinder.conf
#enable_v1_api = true
```

Cinder cli worked.

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -535,3 +539,96 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+
+			sort.Sort(APIVersionsByID(apiversions))
+			// take first CURRENT or SUPPORTED version
+			for _, version := range apiversions {
```
for loop over map is not guarantee to be in order

mkutsevol replied:
apiversions is a []APIVersion, I assumed in order iteration

Can you post the output of your api version query? 
https://developer.openstack.org/api-ref/block-storage/v2/index.html?expanded=list-api-versions-detail#list-api-versions

I'll do unit tests for it.

rootfs replied:
```json
{
    "versions": [
        {
            "status": "SUPPORTED",
            "updated": "2014-06-28T12:20:21Z",
            "id": "v1.0",
            "links": [
                {
                    "href": "http://a.b.c.d:8776/v1/",
                    "rel": "self"
                }
            ]
        },
        {
            "status": "CURRENT",
            "updated": "2012-11-21T11:33:21Z",
            "id": "v2.0",
            "links": [
                {
                    "href": "http://a.b.c.d:8776/v2/",
                    "rel": "self"
                }
            ]
        }
    ]
}
```

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -535,3 +539,96 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+					switch version.ID {
+					case "v2.0":
+						os.bsOpts.BSVersion = "v2"
+						break
```
return from here

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -535,3 +539,96 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+						break
+					case "v1.0":
+						os.bsOpts.BSVersion = "v1"
+						break
```
return from here

rootfs commented:
with the following patch, I verified dynamic provisioning and attach using v2 API
```diff
diff --git a/pkg/cloudprovider/providers/openstack/openstack.go b/pkg/cloudprovider/providers/openstack/openstack.go
index 6ede223..b3b55ab 100644
--- a/pkg/cloudprovider/providers/openstack/openstack.go
+++ b/pkg/cloudprovider/providers/openstack/openstack.go
@@ -593,20 +593,21 @@ func (os *OpenStack) Volumes() (IVolumeService, error) {

                        sort.Sort(APIVersionsByID(apiversions))
                        // take first CURRENT or SUPPORTED version
-                       for _, version := range apiversions {
+                       for i := 0; i < len(apiversions); i++ {
+                               version := apiversions[i]
                                if version.Status == "CURRENT" || version.Status == "SUPPORTED" {
                                        switch version.ID {
                                        case "v2.0":
                                                os.bsOpts.BSVersion = "v2"
-                                               break
+                                               return true, nil
                                        case "v1.0":
                                                os.bsOpts.BSVersion = "v1"
-                                               break
+                                               return true, nil
                                        }
                                }
                        }

-                       return true, nil
+                       return false, fmt.Errorf("no supported block storage API version found")
                })

                if err != nil {
```

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -535,3 +539,96 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
 }
+
+// Implementation of sort interface for blockstorage version probing
+type APIVersionsByID []apiversions_v1.APIVersion
```
add unit test

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -535,3 +539,96 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+}
+
+func (os *OpenStack) volumeService() (volumeService, error) {
+	switch os.bsOpts.BSVersion {
```
unit test this

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -535,3 +539,96 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+			sort.Sort(APIVersionsByID(apiversions))
+			// take first CURRENT or SUPPORTED version
+			for _, version := range apiversions {
+				if version.Status == "CURRENT" || version.Status == "SUPPORTED" {
```
I still think it should only be checking "CURRENT" here (thereby removing the utility of sorting the versions). To me, for anything other than testing, users should explicitly specify the Block Storage (or, indeed, any other service) version they want; if they don't, they should get the current version or an error.

Alternatively, you could check for current versions first, then fall back to supported versions (untested):
```go
for _, status := range []string{"CURRENT", "SUPPORTED"} {
  for _, version := range apiversions {
    if version.Status == status {
	switch version.ID {
	case "v2.0":
		os.bsOpts.BSVersion = "v2"
		return false, nil
	case "v1.0":
		os.bsOpts.BSVersion = "v1"
		return false, nil
	}
    }
  }
}
```

rootfs replied:
SGTM

jrperritt reviewed (changes requested):
Took another pass at the review and left some more feedback.

mkutsevol commented:
@rootfs, @jrperritt thank you very much, will do. 

breaking the switch. niiice. 

gone figuring out how to unit test it.

rootfs commented:
I like comment from @jrperritt on current and support. Choosing current is `prioritized` per [openstack document](https://developer.openstack.org/api-guide/quick-start/):
```text
The API status reflects the state of the endpoint on the service.

Current indicates a stable version that is up-to-date, recent, and might receive future versions. This endpoint should be prioritized over all others.
Supported is a stable version that is available on the server. However, it is not likely the most recent available and might not be updated or might be deprecated at some time in the future.
```

mkutsevol commented:
@rootfs, @jrperritt I've pushed an update, didn't squash so the changes are easily visible. 

Added another config parameter, to enforce 'CURRENT only' in autodetection, by default it is in strict mode and allows only CURRENT api version. 

Autodetection code now has its own func and it is unit tested.

Added error handlers to createvolume, won't panic any more :D

mkutsevol commented:
@k8s-bot verify test this

rootfs commented:
@k8s-bot verify test this

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -555,8 +557,60 @@ func (apiVersions APIVersionsByID) Less(i, j int) bool {
+		}
+	}
+	// Only if we don't support CURRENT api version we get this far
+	glog.Warningf("I don't support the CURRENT blockstorage API version")
```
some better warning message :)

mkutsevol replied:
glog.Warningf("Autoprobe didn't find a suitable CURRENT blockstorage API version") 
?

rootfs replied:
SGTM

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -555,8 +557,60 @@ func (apiVersions APIVersionsByID) Less(i, j int) bool {
+				glog.V(3).Infof("Blockstorage API version probing has found a suitable CURRENT api version: %s", detectedApiVersion)
+				return detectedApiVersion
+			} else {
+				break
```
continue, till `CURRENT` API is found or all versions probed and no `CURRENT` API is found.

mkutsevol replied:
No point in continuing, you cant have 2 "CURRENT" APIs

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -555,8 +557,60 @@ func (apiVersions APIVersionsByID) Less(i, j int) bool {
-func (os *OpenStack) volumeService() (volumeService, error) {
-	switch os.bsOpts.BSVersion {
+func autoVersionSelector(apiVersion *apiversions_v1.APIVersion) string {
+	switch apiVersion.ID {
```
`strings.ToLower(apiVersion.ID)`

mkutsevol replied:
No more believing in OS api returns :)

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -555,8 +557,60 @@ func (apiVersions APIVersionsByID) Less(i, j int) bool {
+	// search for CURRENT api version
+	// There should be only one api version with CURRENT status
+	for _, version := range availableApiVersions {
+		if version.Status == "CURRENT" {
```
`strings.ToUpper(version.Status)`

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -555,8 +557,60 @@ func (apiVersions APIVersionsByID) Less(i, j int) bool {
+		glog.Info("Blockstorage api version autodetection is allowed to search in SUPPORTED versions")
+		sort.Sort(APIVersionsByID(availableApiVersions))
+		for _, version := range availableApiVersions {
+			if version.Status == "SUPPORTED" {
```
`strings.ToUpper(version.Status)`

rootfs commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -555,8 +557,60 @@ func (apiVersions APIVersionsByID) Less(i, j int) bool {
+		}
+	}
+	// Nothing suitable found, failed autodetection
+	errMsg := "Autoprobing for blockstorage API failed to find a suitable api version See auto-allow-supported and bs-version settings."
```
print out `auto-allow-supported` and `bs-version`

mkutsevol replied:
Moved and print settings. But really auto-allow-supported may change bs-version will always be auto

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -89,7 +91,9 @@ type LoadBalancerOpts struct {
 
 type BlockStorageOpts struct {
-	TrustDevicePath bool `gcfg:"trust-device-path"` // See Issue #33128
+	AllowAutoSupported bool   `gcfg:"auto-allow-supported"` // Allow autodetection to fall back to SUPPORTED api versions
```
I would vote against this due to it being too fine-grained. Either just check for "CURRENT" or automatically fallback to "SUPPORTED". If users are clamoring for this feature, it could be added later in a backwards-compatible way.

mkutsevol replied:
easily

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -173,7 +177,9 @@ func readConfig(config io.Reader) (Config, error) {
 	// Set default values for config params
+	cfg.BlockStorage.BSVersion = "auto"
 	cfg.BlockStorage.TrustDevicePath = false
+	cfg.BlockStorage.AllowAutoSupported = false
```
This should default to `false` (without this line). The one above it, as well.

jrperritt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -535,3 +541,127 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
+	}
+}
+
+func doBsApiVersionAutodetect(availableApiVersions []apiversions_v1.APIVersion, AllowAutoSupported bool) string {
```
Instead of duplicating the `for` loop logic, why not just do something like this: https://github.com/kubernetes/kubernetes/pull/40423#discussion_r106489621 ?

mkutsevol replied:
without a a check in the second loop, for sure will do that way

mkutsevol commented:
@jrperritt, pushed

I'll squash, when everyone is happy with it :D

rootfs commented on pkg/cloudprovider/providers/openstack/openstack_test.go:
```diff
@@ -385,3 +391,45 @@ func TestVolumes(t *testing.T) {
+		result_version[ver] = fmt.Sprintf("v%d", ver)
+	}
+	result_version[0] = ""
+	api_current_v1 := apiversions.APIVersion{api_version[1], status_current, updated}
```
check out verify failure, through lines 408-410

mkutsevol replied:
Please say it in anther way, I didn't understand

mkutsevol replied:
@rootfs, Please say it in another way, I didn't understand

rootfs replied:
see 
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/40423/pull-kubernetes-verify/22616/build-log.txt


```text
W0320 09:31:56.138] pkg/cloudprovider/providers/openstack/openstack_test.go:408: k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/blockstorage/v1/apiversions.APIVersion composite literal uses unkeyed fields
W0320 09:31:56.138] pkg/cloudprovider/providers/openstack/openstack_test.go:409: k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/blockstorage/v1/apiversions.APIVersion composite literal uses unkeyed fields
W0320 09:31:56.138] pkg/cloudprovider/providers/openstack/openstack_test.go:410: k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/blockstorage/v1/apiversions.APIVersion composite literal uses unkeyed fields
W0320 09:31:56.147] pkg/cloudprovider/providers/openstack/openstack_test.go:412: k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/blockstorage/v1/apiversions.APIVersion composite literal uses unkeyed fields
W0320 09:31:56.148] pkg/cloudprovider/providers/openstack/openstack_test.go:413: k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/blockstorage/v1/apiversions.APIVersion composite literal uses unkeyed fields
W0320 09:31:56.148] pkg/cloudprovider/providers/openstack/openstack_test.go:415: k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/blockstorage/v1/apiversions.APIVersion composite literal uses unkeyed fields
W0320 09:31:56.148] pkg/cloudprovider/providers/openstack/openstack_test.go:416: k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/blockstorage/v1/apiversions.APIVersion composite literal uses unkeyed fields
```

mkutsevol replied:
Oh! Thank you, @rootfs 
Pushing in an instant.

hogepodge reviewed (approved):
Not to throw a wrench into this discussion, but one thing to consider is the recent release of the Cinder v3 API. The good news is that the initial microversion of v3 is the same API as v2, the bad news is v3 has run ahead with adding new microversioned features which some users might want to take advantage of. I like the autodetect feature, and depending on gophercloud implementation we need to add v3 support. Regardless, I don't think that should hold this patch up, since v2 is the currently supported API version and v1 has been marked as deprecated.

rootfs commented:
@hogepodge 
still waiting v3 api in gophercloud
https://github.com/gophercloud/gophercloud/tree/master/openstack/blockstorage

rootfs commented:
I tested latest code on my setup. I created a storage class, a claim and a Pod. Eventually the claim and pod were created through v2 api.

@jrperritt any comments?

jrperritt commented:
> any comments?

This seems fine to me.

rootfs commented:
@mkutsevol please make another (hopefully last) squash

xsgordon commented:
@hogepodge yes I had been politely ignoring the need for v3 support as while it is ultimately going to be needed the situation with v1 versus v2 is more critical right now. We probably should have a separate issue tracking the need for v3 support eventually though.

mkutsevol commented:
@rootfs, pushed and tests are ok.

idvoretskyi commented:
cc @e0ne

rootfs commented:
/lgtm

rootfs commented:
@wojtek-t can you approve? Thanks.

e0ne commented:
@xsgordon fair enough about v2 and v3. Anyway, from Cinder perspective both v1 and v2 versions are deprecated now. And as was said above, v3 and v2 versions are the same, so IMO it would be good to have v3 support. It will help to add microversions support in the future

xsgordon commented:
@rootfs I couldn't find an open issue for gophercloud V3 support so I filed https://github.com/gophercloud/gophercloud/issues/305 - patches welcome I'm sure ;).

@e0ne I don't disagree that it would be good, but right now we are blocked on the above and in the meantime the situation is that while V2 is deprecated but still supported where as V1 (the only version the provider supports right now) is completely removed, so without this patch the provider's volume integration is not usable at all on OpenStack clouds that use recent releases of OpenStack. V3 support should be added in a separate PR once it is supported via Gophercloud.

hogepodge commented:
Yes please let's unblock on this PR and then work on cinder v3 in gophercloud, then getting it into this patch.

rootfs commented:
@anguslees can you approve?

e0ne commented:
@xsgordon, @hogepodge sure, I won't block this patch anyway.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -86,73 +252,45 @@ func (os *OpenStack) DetachDisk(instanceID string, partialDiskId string) error {
-	if len(disk.Attachments) > 0 && disk.Attachments[0]["server_id"] != nil && instanceID == disk.Attachments[0]["server_id"] {
+	if volume.AttachedServerId != instanceID {
+		errMsg := fmt.Sprintf("Disk: %s has no attachments or is not attached to compute: %s", volume.Name, instanceID)
+		glog.Errorf(errMsg)
```
minor point: This is an openstack-python pattern, and not usual for golang.  Usually you want to glog the error *xor* return the error to your caller, presumably so they might glog it later with some additional context.

mkutsevol replied:
@anguslees, thank you for the review, that is a direct port of logic of previous code. If/when time comes to add v3 support I will keep this in mind.

anguslees commented:
I made a minor style comment, but I'm ok with code being merged as-is.

/approve

wojtek-t commented:
/approve

xrl commented:
How soon does this get in to the 1.5 series?

mkutsevol commented:
@xrl, hi! 
I doubt that it will land into 1.5. It should have severe merge conflicts with 1.5

xrl commented:
OK, so how soon will it land in 1.6? 👍

rootfs commented:
@ethernetdan @calebamiles can this fix be cherry picked into 1.6?
